src/RcppInterface.cpp

```diff
@@ -3,15 +3,74 @@
+// This solution is temporary, and will be changed as soon as the external bug is fixed.
+
 std::vector<std::string> mean_norm::param_names = {"before_mean", "after_mean"};
 std::vector<std::string> var_norm::param_names = {"before_var", "after_var"};
 std::vector<std::string> meanvar_norm::param_names = {"before_mean", "after_mean", " before_var", "after_var"};
 std::vector<std::string> negbin::param_names = {"before_prob", "after_prob"};
 std::vector<std::string> poisson::param_names = {"before_rate", "after_rate"};
 std::vector<std::string> exponential::param_names = {"before_rate", "after_rate"};
 
+std::string mean_norm::factoryName = "mean_norm";
+std::string var_norm::factoryName = "var_norm";
+std::string meanvar_norm::factoryName = "meanvar_norm";
+std::string negbin::factoryName = "negbin";
+std::string poisson::factoryName = "poisson";
+std::string exponential::factoryName = "exponential";
+
+std::string mean_norm::description = "Normal distribution with change in Mean and constant Variance";
+std::string var_norm::description = "Normal Distribution with change in Variance and constant Mean";
+std::string meanvar_norm::description = "Normal distribution with change in both mean and variance";
+std::string negbin::description = "Negative Binomial distribution with change in Probability of Success";
+std::string poisson::description = "Poisson distribution with change in Rate";
+std::string exponential::description = "Exponential distribution with change in Rate";
+
 std::vector<std::string> BS::param_names = {"cpts_index", "cpts", "invalidates_index", "invalidates_after", "cost"};
 
+std::string BS::factoryName = "BS";
+
+std::string BS::description = "Regular Binary Segmentation";
+
+template<>
+std::map<std::string, std::shared_ptr<Distribution>(*)()> GenericFactory<Distribution>::regSpecs =
+        std::map<std::string, std::shared_ptr<Distribution>(*)()>();
+
+template<>
+std::map<std::string, std::shared_ptr<Algorithm>(*)()> GenericFactory<Algorithm>::regSpecs =
+        std::map<std::string, std::shared_ptr<Algorithm>(*)()>();
+
+template<>
+std::map<std::string, std::string> GenericFactory<Algorithm>::regDescs = std::map<std::string, std::string>();
+
+template<>
+std::map<std::string, std::string> GenericFactory<Distribution>::regDescs = std::map<std::string, std::string>();
+
+template<>
+bool Registration<mean_norm, Distribution, DistributionFactory>::is_registered =
+        DistributionFactory::Register(mean_norm::factoryName, mean_norm::description, mean_norm::createMethod);
+template<>
+bool Registration<var_norm, Distribution, DistributionFactory>::is_registered =
+        DistributionFactory::Register(var_norm::factoryName, var_norm::description, var_norm::createMethod);
+template<>
+bool Registration<meanvar_norm, Distribution, DistributionFactory>::is_registered =
+        DistributionFactory::Register(meanvar_norm::factoryName, meanvar_norm::description, meanvar_norm::createMethod);
+template<>
```

Comment:
@diego-urgell same for this code block, please generate it using the macro to avoid repetition

Reply:
@tdhock That was the way it was before, and it was really nice. However, since the C++17 standard is not supported on Windows and Ubuntu for R, I had to remove those features, among them the `inline` initializers. It was possible to initialize those values on the Macro before because of that. But now, it is not possible to do it because of the following reasons: 

1. Cannot use inline because I am stuck with C++11 for now. 
2. Cannot initialize on the constructor, since it is a static variable. 
3. Cannot initialize outside of the class inside the Macro, since those files are included on other source files and the symbols get duplicated, causing a compiler error. 

I really spent several days thinking on how to do this the best possible way while avoiding C++17 features, but I could not come up with a better solution. @rkillick and I agreed that for now it is best to leave it this way. 

After GSOC I will invest more time on this package and then I will try again to use the C++17 features, but I am afraid it has something to do with R (probably a bug when interacting with gcc). If any potential solution comes to mind, please share it so that we can evaluate if it is feasible to execute it now. 

Please see https://github.com/diego-urgell/BinSeg/releases/tag/std_rcpp17

Btw, I am sorry that I merged this branch! I did not realize you had commented until now.

Reply:
I am surprised that you were not able to get it to work with C++ 11 and macros. It seems like that should be possible.
About the C++ 17 not working, you may have stumbled onto a bug with base R package compilation / installation, so it would be helpful for the larger R community if you could follow up with minimal reproducible examples and emails to the lists, as described here https://github.com/diego-urgell/BinSeg/pull/11